src/main/java/ch/njol/skript/expressions/ExprFurnaceTime.java

```diff
@@ -0,0 +1,309 @@
+
+@Name("Furnace Times")
+@Description({
+	"You can use these expressions to change the cook, total cook, and burn time of a furnace.",
+	"If used within any furnace event, you do not have to provide 'of event-block'.",
+	"<ul>",
+	"<li>cook time: The amount of time an item has been smelting for.</li>",
+	"<li>total cook time: The amount of time an item is required to smelt for until done.</li>",
+	"<li>burn time: The amount of time left of the fuel.</li>",
+	"</ul>"
+})
+@Examples({
+	"set the cooking time of {_block} to 10",
+	"set the total cooking time of {_block} to 50",
+	"set the burning time of {_block} to 100",
+	"on smelt:",
+		"\tif the fuel slot is charcoal:",
+			"\t\tadd 5 seconds to the burn time"
+})
+@Since("INSERT VERSION")
+public class ExprFurnaceTime extends PropertyExpression<Block, Timespan> {
+
+	enum FurnaceExpressions {
+		COOKTIME("cook[ing] time"),
+		TOTALCOOKTIME("total cook[ing] time"),
+		BURNTIME("burn[ing] time");
+
+		private String name;
+
+		FurnaceExpressions(String name) {
+			this.name = name;
+		}
+
+	}
+
+	private static final FurnaceExpressions[] furnaceExprs = FurnaceExpressions.values();
+	
+	static {
+
+		int size = furnaceExprs.length;
+		String[] patterns = new String[size * 2];
+		for (FurnaceExpressions value : furnaceExprs) {
+			patterns[2 * value.ordinal()] = "[the] [furnace] " + value.name + " of %blocks%";
+			patterns[2 * value.ordinal() + 1] = "%blocks%['s]" + value.name;
+		}
+
+		Skript.registerExpression(ExprFurnaceTime.class, Timespan.class, ExpressionType.PROPERTY, patterns);
+	}
+
+	private FurnaceExpressions type;
+	private boolean explicitlyBlock;
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		type = furnaceExprs[(int) floor(matchedPattern / 2)];
+		if (exprs[0] != null) {
+			explicitlyBlock = true;
+			setExpr((Expression<Block>) exprs[0]);
+		} else {
+			if (!getParser().isCurrentEvent(FurnaceBurnEvent.class, FurnaceStartSmeltEvent.class, FurnaceExtractEvent.class, FurnaceSmeltEvent.class)) {
+				Skript.error("There's no furnace in a " + getParser().getCurrentEventName() + " event.");
+				return false;
+			}
+			setExpr(new EventValueExpression<>(Block.class));
+		}
+		return true;
+	}
+
+	@Override
+	protected Timespan @Nullable [] get(Event event, Block[] source) {
+		return get(source, new Getter<Timespan, Block>() {
+			@Override
+			public @Nullable Timespan get(Block block) {
+				Furnace furnace = (block != null && block.getState() instanceof Furnace furnaceCheck) ? furnaceCheck : null;
+				switch (type) {
+					case COOKTIME -> {
+						return new Timespan(Timespan.TimePeriod.TICK, (int) furnace.getCookTime());
+					}
+					case TOTALCOOKTIME -> {
+						if (!explicitlyBlock && event instanceof FurnaceStartSmeltEvent startEvent) {
+							return new Timespan(Timespan.TimePeriod.TICK, startEvent.getTotalCookTime());
+						} else {
```

Comment:
this concerns me, since `time of event-block` and `time` might return different values.

Reply:
Should I test that out or do you have something in mind?

Reply:
you should definitely test that